File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-2.java
class Solution {
    public int longestIncreasingPath(int[][] matrix) {
        int[][] dp = new int[matrix.length + 1][matrix[0].length + 1];
        for (int[] arr : dp) Arrays.fill(arr, -1);
        int max = 0;
        for (int r = 0; r < matrix.length; r++) {
            for (int c = 0; c < matrix[0].length; c++) {
                max = Math.max(max, dfs(dp, matrix, r, c) + 1);
            }
        }
        return max;
    }

    int[][] directions = {
        {1, 0}, {-1, 0},
        {0, 1}, {0, -1}
    };

    public int dfs(int[][] dp, int[][] matrix, int r, int c) {
        if (dp[r][c] != -1) return dp[r][c];
        int max = 0;
        for (int[] direction : directions) {
            int nR = r + direction[0];
            int nC = c + direction[1];
            if (nR < 0 || nR >= matrix.length || nC < 0 || nC >= matrix[0].length) continue;
            if (matrix[nR][nC] <= matrix[r][c]) continue;
            max = Math.max(max, dfs(dp, matrix, nR, nC) + 1);
        }
        dp[r][c] = max;
        return dp[r][c];
    }
}

File: Data Structures & Algorithms/longest-common-subsequence/submission-2.cpp
class Solution {
public:

    int longestCommonSubsequence(string text1, string text2) {
        vector<vector<int>> dp(text1.size(), vector<int>(text2.size(), -1));
        return dfs(dp, text1, text2, 0, 0);
    }

    int dfs(vector<vector<int>>& dp, string& text1, string& text2, int i, int j) {
        if (i == text1.size() || j == text2.size()) return 0;
        if (dp[i][j] != -1) return dp[i][j];
        int max = -1;
        if (text1[i] == text2[j]) {
            max = dfs(dp, text1, text2, i + 1, j + 1) + 1;
        }
        max = std::max(max, dfs(dp, text1, text2, i, j + 1));
        max = std::max(max, dfs(dp, text1, text2, i + 1, j));
        dp[i][j] = max;
        return dp[i][j];
    }
};
